invoke-ai/InvokeAI
# API model downloads are now threaded and generate progress events

# Review not needed on this PR

I'm not going to use this code, but want to keep it around as an example to refer to as I redo the idea correctly.


## What type of PR is this? (check all applicable)

- [X ] Feature
- [ X] Optimization


## Have you discussed this change with the InvokeAI team?
- [ X] Yes

      
## Description

Currently, when the WebGUI's Model Manager is used to download a model via its remote URL or repo_id, the whole frontend freezes until the download is finished. This PR starts a thread to download the model in the background. While the model is downloading, a series of events is fired:

* `model_import_started(import_path:str)`
* `download_started(url:str)`
* `download_progress(url:str, downloaded_bytes:int, total_bytes:int)`
* `download_completed(url:str, status_code:int, download_path:str)`
* `model_import_completed(import_path:str, import_info:AddModelResults, success:bool, error:str)`

**download_progress** events are generated on every 10 MB of data downloaded. Unsuccessful import jobs are reported with a `model_import_completed` event with `success=False` and an error message in `error`.

I ran into a lot of circular import errors while writing this, and as a result several methods that accept `EventServiceBase` objects use an untyped argument for the event object.

## Related Tickets & Documents

See discord https://discord.com/channels/1020123559063990373/1128608407864217630/1130217223651336303


## QA Instructions, Screenshots, Recordings

After launching the web server, go to http://localhost:9090/doc and look for the `Import Model /api/v1/models/import` API call. Try installing a civitai model (or any URL or repo_id) using this body:
```
{
  "location": "http://civitai.com/...",
  "prediction_type": null
}
```

You should see a 200 response of "request queued" and console log messages indicating that the import/download events are being issued.



## Added/updated tests?

- [X] No : Hard to test this type of API.
